cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_kernel_hist_impl_dpc.hpp

```diff
@@ -0,0 +1,2966 @@
+                }
+                else {
+                    level_node_lists.push_back(node_list_new);
+
+                    node_vs_tree_map_list = node_vs_tree_map_list_new;
+
+                    last_event = train_service_kernels_.do_level_partition_by_groups(
+                        ctx,
+                        full_data_nd_,
+                        node_list,
+                        tree_order_lev_,
+                        tree_order_lev_buf_,
+                        ctx.row_count_,
+                        ctx.selected_row_total_count_,
+                        ctx.column_count_,
+                        node_count,
+                        ctx.tree_in_block_,
+                        { last_event });
+                }
+            }
+
+            node_count = node_count_new;
+        }
+
+        model_manager.add_tree_block(level_records, bin_borders_host_, iter_tree_count);
+
+        for (Index tree_idx = 0; tree_idx < iter_tree_count; ++tree_idx) {
+            compute_results(ctx,
+                            model_manager,
+                            data_host_,
+                            response_host_,
+                            oob_rows_list,
+                            oob_row_count_list,
+                            oob_per_obs_list_,
+                            res_var_imp_,
+                            var_imp_variance_host_,
+                            engine_arr,
+                            tree_idx,
+                            iter_tree_count,
+                            iter,
+                            { last_event })
+                .wait_and_throw();
+        }
+    }
+
+    // Finalize results
+    if (ctx.oob_err_required_ || ctx.oob_err_obs_required_) {
+        pr::ndarray<Float, 1> res_oob_err;
+        pr::ndarray<Float, 1> res_oob_err_obs;
+
+        finalize_oob_error(ctx, response_host_, oob_per_obs_list_, res_oob_err, res_oob_err_obs)
+            .wait_and_throw();
+
+        if (ctx.oob_err_required_) {
+            auto res_oob_err_host = res_oob_err.to_host(queue_);
+            res.set_oob_err(homogen_table::wrap(res_oob_err_host.flatten(), 1, 1));
+        }
+
+        if (ctx.oob_err_obs_required_) {
+            auto res_oob_err_obs_host = res_oob_err_obs.to_host(queue_);
+            res.set_oob_err_per_observation(
+                homogen_table::wrap(res_oob_err_obs_host.flatten(), ctx.row_count_, 1));
+        }
+    }
+
+    if (ctx.mdi_required_ || ctx.mda_required_) {
+        finalize_var_imp(ctx, res_var_imp_, var_imp_variance_host_).wait_and_throw();
+        auto res_var_imp_host = res_var_imp_.to_host(queue_);
+        res.set_var_importance(
+            homogen_table::wrap(res_var_imp_host.flatten(), 1, ctx.column_count_));
+    }
+
+    return res.set_model(model_manager.get_model());
+}
+
+#define INSTANTIATE(F, B, I, T) template class train_kernel_hist_impl<F, B, I, T>;
+
+} // namespace oneapi::dal::decision_forest::backend
+
+#endif
```

Comment:
Too big file for me

Reply:
will think regarding how to split it for next MR